File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/welcome/WelcomePageAdapter.java
package openfoodfacts.github.scrachx.openfood.features.welcome;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.StringRes;
import androidx.viewpager.widget.PagerAdapter;

public class WelcomePageAdapter extends PagerAdapter {
    private final LayoutInflater layoutInflater;
    private final int[] layouts;

    public WelcomePageAdapter(LayoutInflater layoutInflater, @StringRes int[] layouts) {
        this.layoutInflater = layoutInflater;
        this.layouts = layouts;
    }

    @NonNull
    @Override
    public Object instantiateItem(@NonNull ViewGroup container, int position) {
        View view = layoutInflater.inflate(layouts[position], container, false);
        container.addView(view);

        return view;
    }

    @Override
    public int getCount() {
        return layouts.length;
    }

    @Override
    public boolean isViewFromObject(@NonNull View view, @NonNull Object obj) {
        return view == obj;
    }

    @Override
    public void destroyItem(@NonNull ViewGroup container, int position, @NonNull Object object) {
        View view = (View) object;
        container.removeView(view);
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/welcome/WelcomePageAdapter.kt
package openfoodfacts.github.scrachx.openfood.features.welcome

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.viewpager.widget.PagerAdapter

class WelcomePageAdapter(
        private val layoutInflater: LayoutInflater,
) : PagerAdapter() {
    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val layout = WelcomeScreen.values()[position].layout
        val view = layoutInflater.inflate(layout, container, false)
        container.addView(view)
        return view
    }

    override fun getCount() = WelcomeScreen.values().size

    override fun isViewFromObject(view: View, obj: Any) = view === obj

    override fun destroyItem(container: ViewGroup, position: Int, item: Any) = container.removeView(item as View)
}